Improving shots estimate to give errors within user-specified precision.

jsaroni opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
Fixes #issue-12414 (https://github.com/Qiskit/qiskit/issues/12414). When the user specifies the precision in EstimatorV2.run, the actual standard deviation from results is not bounded by or close to the user-specified precision for an observable that is a general linear combination of Pauli strings. This pull request gives a more accurate approximation to the number of shots needed for the desired precision for such observables. The tests to verify this are based on https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.EstimatorV2 and https://github.com/Qiskit/qiskit-ibm-runtime/issues/1684. 

### Details and comments
The shots estimation accounts for the sum of the absolute value squared of the coefficients of the observable as proposed in https://github.com/Qiskit/qiskit/issues/12414 to give a standard deviation that is bounded by and closer to the user-specified precision for any observable. In backend_estimator_v2.py, the _run function is modified to include the sum of the absolute value squared of the coefficients of the observable in the shots estimation.

The following plot shows that the ratio of precision from results to the user-specified precision changes with varying observable term coefficients using the current implementation. In the proposed implementation, this ratio is fixed and close to 1. 
![average_precision_over_runs](https://github.com/Qiskit/qiskit/assets/77505813/0c89fa9b-1ece-43c8-a22b-dcdbc083f570)
The plot below shows that in the current implementation, having more than one occurrence of a term in an observable leads to precision results that are different from the user-specified precision due to a fixed number of shots for observables with different numbers of terms and error propagation. In the proposed implementation, the precision from the results remains close to the user-specified precision.
![ZZ_linear_combinations](https://github.com/Qiskit/qiskit/assets/77505813/c9fb1c58-abb7-40b7-bdea-09e0282ce3a4)

jsaroni committed:
Update backend_estimator_v2.py

Improving shot estimate to give errors within user-given precision.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=12618) <br/>All committers have signed the CLA.

ihincks commented on qiskit/primitives/backend_estimator_v2.py:
```diff
@@ -160,9 +160,15 @@ def _validate_pubs(self, pubs: list[EstimatorPub]):
+        bc_obs = data.observables
+        coeff_sq_sum = 0
+        for pauli, coeff in bc_obs[0].items():
+            coeff_sq_sum += coeff**2
```
This has an easy 1-line replacement. You're also not using `pauli`, so it shouldn't be part of the iteration.

```suggestion
        coeff_sq_sum = sum(coeff**2 for coeff in bc_obs[0].values())
```

jsaroni replied:
Thanks. I included test plots and made a commit to include a test function in qiskit/test/python/primitives/test_backend_estimator_v2.py. The tests are passed by the proposed implementation but not the current. Also used the 1-line replacement and removed pauli.

ihincks commented on qiskit/primitives/backend_estimator_v2.py:
```diff
@@ -160,9 +160,15 @@ def _validate_pubs(self, pubs: list[EstimatorPub]):
         pub_dict = defaultdict(list)
+        for pub in pubs:
+            data = self._preprocess_pub(pub)
+        bc_obs = data.observables
```
You're preprocessing all of the pubs, but then just looking at the last one.

jsaroni replied:
Right. I fixed this and checked that the code works when there is more than one pub in test_backend_estimator_v2.py test_error.

ihincks commented on qiskit/primitives/backend_estimator_v2.py:
```diff
@@ -160,9 +160,15 @@ def _validate_pubs(self, pubs: list[EstimatorPub]):
         # consolidate pubs with the same number of shots
         for i, pub in enumerate(pubs):
-            shots = int(math.ceil(1.0 / pub.precision**2))
+            shots = int(math.ceil(coeff_sq_sum / pub.precision**2))
```
Why not make coeff_sq_sum depend on the pub? With what you've written, it only depends on the last pub.

Moreover, it only depends on the first element of the last pub.

jsaroni replied:
Sounds good. I guess it was indirectly dependent it. 

From the EstimatorV2 tutorial example bc_obs is of the form [{'II': 1.0, 'IZ': 2.0, 'XI': 3.0}] so bc_obs[0] gets all the dictionary terms {'II': 1.0, 'IZ': 2.0, 'XI': 3.0}.

ihincks commented on qiskit/primitives/backend_estimator_v2.py:
```diff
@@ -160,9 +160,15 @@ def _validate_pubs(self, pubs: list[EstimatorPub]):
     def _run(self, pubs: list[EstimatorPub]) -> PrimitiveResult[PubResult]:
         pub_dict = defaultdict(list)
+        for pub in pubs:
+            data = self._preprocess_pub(pub)
```
Please look through the `_preprocess_pub` function. It doesn't do anything relevant to your changes, so it seems like a waste to call it at all.

jsaroni replied:
I had to use _preprocess_pub to get the right data type for using .values().

ihincks reviewed (commented):
Thanks for your submission.

Part of this PR should be evidence that it indeed helps the situation. This could be a plot commented on the PR, exploring different regimes. It should also include tests in the test folder.

jsaroni committed:
Update backend_estimator_v2.py

Used 1-line replacement for coeff_sq_sum removing pauli. Had to use _preprocess_pub to get the right data type for using .values(). From the EstimatorV2 tutorial example bc_obs is of the form [{'II': 1.0, 'IZ': 2.0, 'XI': 3.0}] so bc_obs[0] gets all the dictionary terms {'II': 1.0, 'IZ': 2.0, 'XI': 3.0}.

jsaroni committed:
Update test_backend_estimator_v2.py

Included a function test_error which shows that for the proposed shot implementation, the user-specified precision is close to the precision from results within the set tolerance by passing the tests. The tests are not passed by the current implementation.

jsaroni committed:
Merge branch 'main' into main

ihincks commented on qiskit/primitives/backend_estimator_v2.py:
```diff
@@ -162,7 +162,10 @@ def _run(self, pubs: list[EstimatorPub]) -> PrimitiveResult[PubResult]:
         for i, pub in enumerate(pubs):
-            shots = int(math.ceil(1.0 / pub.precision**2))
+            data = self._preprocess_pub(pub)
+            bc_obs = data.observables
```
Why not?
```suggestion
            bc_obs = pub.observables
```
It looks like you're doing a binding and then throwing it away. Why can't you just get the observables directly?

ihincks commented on qiskit/primitives/backend_estimator_v2.py:
```diff
@@ -162,7 +162,10 @@ def _run(self, pubs: list[EstimatorPub]) -> PrimitiveResult[PubResult]:
-            shots = int(math.ceil(1.0 / pub.precision**2))
+            data = self._preprocess_pub(pub)
+            bc_obs = data.observables
+            coeff_sq_sum = sum(coeff**2 for coeff in bc_obs[0].values())
```
Why are we looking only at the first observable? Don't we want to look at the worst case or something? Eg, consider the case

```
observables = [{"XX": 1}, {"YY": 100}]
```

Since we only look at the first observable, we'll be way off for the second one.

ihincks replied:
Eg you could try
```python
        for i, pub in enumerate(pubs):
            coeff_sq_sum = 0
            for obs in pub.observables.ravel():
                coeff_sq_sum = max(coeff_sq_sum, sum(coeff**2 for coeff in obs.values()))
            shots = int(math.ceil(coeff_sq_sum / pub.precision**2))
            pub_dict[shots].append(i)
```

jsaroni replied:
I tried this and it worked, thanks.

ihincks commented on test/python/primitives/test_backend_estimator_v2.py:
```diff
@@ -388,6 +388,24 @@ def test_diff_precision(self, backend, abelian_grouping):
+        job = estimator.run([(psi1, hamiltonian1, [theta1]), (psi1, hamiltonian1, [theta1])], precision=self._precision)
+        result = job.result()
+        np.testing.assert_allclose(result[0].data.stds, [self._precision], rtol=self._rtol)
+        np.testing.assert_allclose(result[1].data.stds, [self._precision], rtol=self._rtol)
```
This should also test observable arrays that aren't 1D.

Also, since your change is specifically about the role of precision in the context of multi-term observables, you should define your own observables here, and not trust that hamiltonian1 is in the format that you like; someone might change it later and not read this test.

jsaroni replied:
Ok.

I defined my own observables including multi-term observables.

ihincks commented on test/python/primitives/test_backend_estimator_v2.py:
```diff
@@ -388,6 +388,24 @@ def test_diff_precision(self, backend, abelian_grouping):
 
+    @combine(backend=BACKENDS, abelian_grouping=[True, False])
+    def test_error(self, backend, abelian_grouping):
+        "Test for error"
```
This name is a bit too generic because `error` can mean many things. Consider renaming to something that includes the word "precision". The string describing the test is also a bit too generic.

jsaroni replied:
Changed test function name to test_precision_error.

ihincks reviewed (commented):
Thanks @jsaroni! I've left a few comments.

jsaroni committed:
Update backend_estimator_v2.py

Using ravel instead of _preprocess_pub to get coefficients directly as shown by @ihincks

jsaroni committed:
Update test_backend_estimator_v2.py

Changed test function name to test_precision_error and defined multi-term observables to test precision.

jsaroni committed:
Merge branch 'main' into main

jsaroni committed:
Merge branch 'Qiskit:main' into estimator_primitive

jsaroni committed:
Merge branch 'Qiskit:main' into main

jsaroni committed:
Merge branch 'estimator_primitive'

jsaroni commented:
Closed in favor of #12711
